pghoard/transfer.py

```diff
@@ -99,15 +101,161 @@ def operation(self):
 TransferQueue = Queue
 
 
+@dataclass
+class TransferIncrement:
+    n_bytes: float
+    tracked_at: float = dataclasses.field(default_factory=time.monotonic)
+
+
+@dataclass
+class UploadEventProgress:
+    key: str
+    file_size: Optional[int]
+    file_type: FileType
+    increments: List[TransferIncrement] = dataclasses.field(default_factory=list)
+    started_at: float = dataclasses.field(default_factory=time.monotonic)
+
+    def is_completed(self) -> bool:
+        # UploadEvents without file_size are allowed, therefore we cannot determine if the event is completed
+        if not self.file_size:
+            return False
+
+        total_nbytes = sum(increment.n_bytes for increment in self.increments)
+        return total_nbytes >= self.file_size
+
+
+class UploadEventProgressTracker(PGHoardThread):
+    CHECK_FREQUENCY: int = 5  # check every 5 seconds for progress
+    WARNING_TIMEOUT: int = 5 * 60  # log a warning in case there is no progress during last 5 minutes
+
+    def __init__(self, metrics: Metrics) -> None:
+        self.metrics = metrics
+        self.log = logging.getLogger("UploadEventProgressTracker")
+
+        self.running: bool = False
+
+        self._tracked_events: Dict[str, UploadEventProgress] = {}
+        self._tracked_events_lock = threading.Lock()
+        self.log.debug("UploadEventProgressTracker initialized")
+
+        super().__init__()
+
+    def track_upload_event(self, file_key: str, file_type: FileType, file_size: Optional[int]) -> None:
+        with self._tracked_events_lock:
+            self.log.info(f"Tracking upload event for file {file_key}")
+            self._tracked_events[file_key] = UploadEventProgress(
+                key=file_key,
+                file_type=file_type,
+                file_size=file_size
+            )
+
+    def untrack_upload_event(self, file_key: str) -> None:
+        if file_key not in self._tracked_events:
+            return
+
+        with self._tracked_events_lock:
+            self._tracked_events.pop(file_key)
+
+    def increment(self, file_key: str, n_bytes: float) -> None:
+        with self._tracked_events_lock:
+            if file_key not in self._tracked_events:
+                raise Exception(f"UploadEvent for {file_key} is not being tracked.")
```

Comment:
**comment** Maybe should raise something like "UntrackedUploadEventError", otherwise this will stop the entire thread

Reply:
ah no, should be alright